📖 Windows Support CAEP (#3616)
kubernetes-sigs/cluster-api: docs/proposals/20200804-windows-support.md

      - [As an operator, I would like to manage Windows OS worker nodes with the CAPI](#as-an-operator-i-would-like-to-manage-windows-os-worker-nodes-with-the-capi)
    - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
      - [Signing of the components.](#signing-of-the-components)
      - [Known prototypes and prior work:](#known-prototypes-and-prior-work)
    - [Security Model](#security-model)
    - [Risks and Mitigations](#risks-and-mitigations)
  - [Alternatives](#alternatives)
  - [Upgrade Strategy](#upgrade-strategy)
  - [Additional Details](#additional-details)
    - [Test Plan [optional]](#test-plan-optional)
    - [Graduation Criteria [optional]](#graduation-criteria-optional)
      - [Alpha](#alpha)
      - [Beta](#beta)
      - [Stable](#stable)
    - [Version Skew Strategy](#version-skew-strategy)
  - [Implementation History](#implementation-history)

## Glossary

Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).

If this proposal adds new terms, or defines some, make the changes to the book's glossary when in PR stage.

## Summary

> This proposal is for the support of the Windows nodes with CAPI and [infrastructure providers](https://cluster-api.sigs.k8s.io/reference/glossary.html#infrastructure-provider) that wish to support Windows [OS](https://cluster-api.sigs.k8s.io/reference/glossary.html#operating-system) worker nodes.  Windows has been Stable in Kubernetes since 1.14 and is supported in mix worker node scenarios.  Windows support will provide the ability to add Windows nodes to a [workload cluster](https://cluster-api.sigs.k8s.io/reference/glossary.html#workload-cluster).  Windows nodes are configured differently across solutions and the centralization of the scripts and guidance in the Cluster API solution will provide guidance for the community on best practices for configuring Windows nodes at scale.

Author: Not sure if there was a good way to do it but I updated to split longer lines.  Hope this helps.

Windows node support has some unique challenges because of the current limitations of Windows Containers.  Windows containers do not have privileged containers which means that configuration and access to the host machine must be done at provisioning time.  An example of this is kube-proxy which is typically run as a Windows service on the host machine since it cannot be deployed as a Daemon set as it is on Linux.  To address this limitation the community has built tools such as the [CSI-Proxy](https://github.com/kubernetes-csi/csi-proxy), which is a CSI driver specific proxy. This proposal will address how to approach the configuration of components that are typically deployed as daemon sets when bootstrapping Windows nodes in CAPI.

## Motivation
The motivation of this project is to establish a set of scripts that will allow cloud providers to bootstrap Windows nodes in Cluster API using best practices. This will enable customers to deploy mixed-OS workloads.

### Goals

> - Enable the creation and management Windows worker nodes on workload clusters through CAPI and infrastructure providers

Reviewer: ```suggestion
- Enable the creation and management of Windows worker nodes on workload clusters through CAPI and infrastructure providers
```

> - Provide community guidance and scripts for building base images with the use of windows nodes

Reviewer: I assume it would make sense to consistently capitalize Windows throughout this document?

- Re-use of the existing Cluster API Bootstrap Provider Kubeadm and other tools where appropriate

### Non-Goals/Future Work

- Provide a way to run [control plane](https://cluster-api.sigs.k8s.io/reference/glossary.html#control-plane) nodes as windows

> - Support for Windows versions outside of the kubernetes support versions

Reviewer: Please capitalize Kubernetes throughout

Author: Updated.Thanks!

- Support for Windows nodes on the [management](https://cluster-api.sigs.k8s.io/reference/glossary.html#management-cluster) or [bootstrap clusters](https://cluster-api.sigs.k8s.io/reference/glossary.html#bootstrap-cluster)

## Proposal

> There several components that need to be addressed for Windows support:

Reviewer: ```suggestion
There are several components that need to be addressed for Windows support:
```

Cloud-init does not have Windows support. An alternative product is [cloudbase-init](https://github.com/cloudbase/cloudbase-init). Cloudbase-init functions in the same way as cloud-init and can consume cloud-init scripts as provided by the Cluster API Bootstrap Provider Kubeadm.  By using cloudbase-init, Windows can leverage the existing solutions and stay up to date with the latest changes in CABPK.

#### VHD Creation
Using cloudbase-init requires the creation of a VHD with the tooling installed on it since it is not provided out of the box by any cloud providers.  To address this we will provide packer scripts as part of the [image-builder project](https://github.com/kubernetes-sigs/image-builder) that will pre-install cloudbase init.  It is important to note that while scripts can be provided to build a  VHD.  All VHD’s built will need to adhere to [Windows licensing requirements](https://www.microsoft.com/en-us/licensing/product-licensing/windows-server).

AKS-Engine has an example implementation for using packer and scripts to do VHD configuration: https://github.com/Azure/aks-engine/blob/master/vhd/packer/windows-vhd-builder.json

> Although the Linux implementation in image-builder uses Ansible for configuration, Windows will not share the configuration. Therefore, Ansible is not a requirement for Windows image builder implementation.  The main requirement for VHD building is that the build can be started from a Linux or Windows machine.

Author: I've started on https://github.com/kubernetes-sigs/image-builder/issues/30 and starting to lean towards using Ansible  due some heavy customization in the scripts.  Anyone have thoughts on this?

Author: see https://github.com/kubernetes-sigs/image-builder/pull/382

#### netbios names
Netbios name resolution concerns were addressed in https://github.com/kubernetes-sigs/cluster-api/issues/2217#issuecomment-598999436

### Infrastructure provider implementation
By leveraging cloudbase-init, an infrastructure provider implementation will require only a few changes which include:

- Make changes to their provider api to enable Windows OS worker nodes 
- Ensuring the cloudbase-init data lands in the correct place on the machines for Windows OS worker nodes

From the infrastructure provider perspective, there are no known required changes to the CAPI API to support Windows nodes at this time.  If during alpha we identify changes we will open issues pertaining to the changes required.

### User Stories

#### As an operator, I would like to create Windows OS worker nodes with the CAPI API.

> #### As an operator, I would like to manage Windows OS worker nodes with the CAPI 

Reviewer: missing API?

### Implementation Details/Notes/Constraints

Due to the lack of privileged containers in Windows there are two options for configuring the components such as kube-proxy, kubelet.  The above solution using wins is preferred because it makes the move to privileged containers straightforward as a drop and replace. 

While this is the best choice for the alpha and the community direction there are some infrastructure providers that may not be able to use wins due to signing or security concerns since wins allows the execution of any arbitrary command on the host. Pre/post commands can be used as an alternative with additional scripts cached on the VHD that enable the configuration.

#### Signing of the components.  
Some infrastructure providers will require any scripts and binaries are signed before deployment.  This will be managed by providing the ability to provide external scripts and binaries during the VHD building process.  We will also be discussing this further with the Image Builder repository team during the process for adding the VHDs.

#### Known prototypes and prior work: 
- https://github.com/adelina-t/cloudbase-init-capz-demo

> - https://github.com/benmoss/kubeadm-windows/tree/master/cluster-api-aws

Author: https://github.com/microsoft/cluster-api-provider-azurestackhci was just opened sourced which has windows components

### Security Model

Wins is currently recommended way to use kubeadm.  Limiting access to the named pipes that are required is one way to mitigate access.   Wins is currently required during and after provisioning for running kube-proxy and the CNI daemonset.  

> The security model for Privileged containers is still being discussed in KEP and is still early and will be address by CAPI when the KEP progresses.

Reviewer: ```suggestion
The security model for Privileged containers is still being discussed in KEP and is still early and will be addressed by CAPI when the KEP progresses.
```

> Kubeadm bootstrap token should be able to use multi-part mime documents for cloudbase-init as done for [Linux in CAPA](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/28d01d064cc2e5b0286ae23b3be7203f18b00447/controllers/awsmachine_controller.go#L601). This will require an update to Cloudbase-init which does support [mutli-part mime documents](https://cloudbase-init.readthedocs.io/en/latest/userdata.html#multi-part-content) but is missing the [boothook](https://cloudinit.readthedocs.io/en/latest/topics/format.html?highlight=boothook#cloud-boothook) functionality.  Support for cloudhooks will should be added to cloudbase-init to meet the AWS provider requirement that would bring parity to Linux implementation during the Beta phase.

Reviewer: "will should"

---
title: Windows Support
authors:
  - "@jsturtevant"
  - "@ksubrmnn"
reviewers:

>   - ""

Author: do they need the be reviewers inside cluster-api or can they be members of the kubernetes-sigs?

>   - "@ncdc"

Reviewer 2: let's also add @randomvariable

    - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
      - [Signing of the components.](#signing-of-the-components)
      - [Known prototypes and prior work:](#known-prototypes-and-prior-work)
    - [Security Model](#security-model)
    - [Risks and Mitigations](#risks-and-mitigations)
  - [Alternatives](#alternatives)
  - [Upgrade Strategy](#upgrade-strategy)
  - [Additional Details](#additional-details)
    - [Test Plan [optional]](#test-plan-optional)
    - [Graduation Criteria [optional]](#graduation-criteria-optional)
      - [Alpha](#alpha)
      - [Beta](#beta)
      - [Stable](#stable)
    - [Version Skew Strategy](#version-skew-strategy)
  - [Implementation History](#implementation-history)

## Glossary

Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).

## Summary

This proposal is for the support of the Windows nodes with CAPI and [infrastructure providers](https://cluster-api.sigs.k8s.io/reference/glossary.html#infrastructure-provider) that wish to support Windows [OS](https://cluster-api.sigs.k8s.io/reference/glossary.html#operating-system) worker nodes.  

> Windows has been Stable in Kubernetes since 1.14 and is supported in mix worker node scenarios.  

Author: updated the wording

Windows support will provide the ability to add Windows nodes to a [workload cluster](https://cluster-api.sigs.k8s.io/reference/glossary.html#workload-cluster).  Windows nodes are configured differently across 
solutions and the centralization of the scripts and guidance in the Cluster API solution will provide 

> guidance for the community on best practices for configuring Windows nodes at scale.

Author: The point I was trying to make is covered in the Goals section and I don't think this is required. Will remove.

guidance for the community on best practices for configuring Windows nodes at scale.

Windows node support has some unique challenges because of the current limitations of Windows Containers. 
Windows containers do not have privileged containers which means that configuration and access to the host

> machine must be done at provisioning time.  

Reviewer 3: Nit: Maybe reword to either 'Windows containers do not support privileged operations' or 'Windows does not support privileged containers'.
Windows containers do not have privileged containers is a bit confusing to me.

> An example of this limitation is how kube-proxy gets configured on Windows nodes. Kube-proxy typically run as a 

Reviewer 4: ```suggestion
An example of this limitation is how kube-proxy gets configured on Windows nodes. Kube-proxy typically runs as a 
```

> An example of this is kube-proxy which is typically run as a Windows service on the host machine since it

Reviewer 3: ```suggestion
An example of this limitation is how kube-proxy gets configured on Windows nodes. Kube-proxy typically run as a Windows service on the host machine and it
```

> Windows service on the host machine and it cannot be deployed as a Daemon set as it is on Linux.  

Reviewer 4: ```suggestion
Windows service on the host machine and it cannot be deployed as a DaemonSet as it is on Linux.  
```

cannot be deployed as a Daemon set as it is on Linux.  To address this limitation the community has built
tools such as the [CSI-Proxy](https://github.com/kubernetes-csi/csi-proxy), which is a CSI driver specific
proxy. This proposal will address how to approach the configuration of components that are typically
deployed as daemon sets when bootstrapping Windows nodes in CAPI.

## Motivation

> The motivation of this project is to establish a set of scripts that will allow cloud providers to

Reviewer 3: +1

> Kubernetes has supported Windows workloads since the release of Windows support in Kubernetes 1.14. The motivation of this project is to enable Cluster API customers to deploy Windows as part of a mixed OS cluster via the Cluster API automation built for platform operators.  This will enable cluster operators to define Windows machines in the same consistent and repeatable fashion.

Reviewer 4: ```suggestion
Kubernetes has supported Windows workloads since the release of Windows support in Kubernetes 1.14. The motivation of this proposal is to enable Cluster API users to deploy Windows as part of a mixed OS cluster via the Cluster API automation built for platform operators.  This will enable cluster operators to define Windows machines in the same consistent and repeatable fashion.
```

bootstrap Windows nodes in Cluster API using best practices. This will enable customers to deploy mixed-OS workloads.

### Goals

- Enable the creation and management of Windows worker nodes on workload clusters through CAPI and infrastructure providers
- Provide community guidance and scripts for building base images with the use of Windows nodes
- Re-use of the existing Cluster API Bootstrap Provider Kubeadm and other tools where appropriate

> ### Non-Goals/Future Work

Reviewer 3: Would it be good to call out that updates to kubeadm needed to for configuration are out-of-scope for this document (such as configuring the node with containerd)?

Author: Configuration of containerd is outside the scope of kubeadm. 

If we find a bug in Kubeadm that stops windows nodes from joining (it is in beta now and I haven't seen any major issues), I think it would need to resolved but I don't see any need to push features into kubeadm at this point and is outside the scope.  Is that what your referring to?

Author: Can this be closed?

Author: going to mark resolved, If still an issue please raise again

- Provide a way to run [control plane](https://cluster-api.sigs.k8s.io/reference/glossary.html#control-plane) nodes as Windows
- Support for Windows versions outside of the Kubernetes support versions

> - Support for Windows nodes on the [management](https://cluster-api.sigs.k8s.io/reference/glossary.html#management-cluster) or [bootstrap clusters](https://cluster-api.sigs.k8s.io/reference/glossary.html#bootstrap-cluster)

Reviewer 2: Consider adding supporting windows nodes with other/non-kubeadm based bootstrap providers as non-goal/future work

There are several components that need to be addressed for Windows support:

- [Cluster API Bootstrap Provider Kubeadm](https://github.com/kubernetes-sigs/cluster-api/tree/master/bootstrap/kubeadm)
- Image creation 
- Kubelet and other component configuration
- Infrastructure provider implementation 

### Cluster API Bootstrap Provider Kubeadm

#### cloud-init and cloudbase-init
For Linux, when using the Kubeadm bootstrap provider, the bootstrap script is provided to the infrastructure provider as a cloud-init script. 

> The infrastructure provider is responsible for putting the cloud-init script in the right location. 

Author: I've learned a bit more about how cloudbase-init finds the cloud-init script in the userdata and it requires configuration of cloudbase-init metadata services. I've update the wording here to reflect that and included a sample.

When the VM is booted, the cloud-init script runs automatically. 

Cloud-init does not have Windows support. An alternative product is [cloudbase-init](https://github.com/cloudbase/cloudbase-init). 
Cloudbase-init functions in the same way as cloud-init and can consume cloud-init scripts as provided by the Cluster API Bootstrap Provider Kubeadm.  

> By using cloudbase-init, Windows can leverage the existing solutions and stay up to date with the latest changes in CABPK.

Author: There is no specific documentation on feature gaps. I will add a link to the documentation for reference.

#### Image Creation

> Using cloudbase-init requires the creation of a image with the tooling installed on it since it is not 

Reviewer 4: ```suggestion
Using cloudbase-init requires the creation of an image with the tooling installed on it since it is not 
```

#### VHD Creation
Using cloudbase-init requires the creation of a VHD with the tooling installed on it since it is not 
provided out of the box by any cloud providers.  To address this we will provide packer scripts as part of 
the [image-builder project](https://github.com/kubernetes-sigs/image-builder) that will pre-install 
cloudbase init.  It is important to note that while scripts can be provided to build a  VHD.  All VHD’s 
built will need to adhere to [Windows licensing requirements](https://www.microsoft.com/en-us/licensing/product-licensing/windows-server).

AKS-Engine has an example implementation for using packer and scripts to do VHD configuration: https://github.com/Azure/aks-engine/blob/master/vhd/packer/windows-vhd-builder.json

Although the Linux implementation in image-builder uses Ansible for configuration, Windows will not share 
the configuration. Therefore, Ansible is not a requirement for Windows image builder implementation.  
The main requirement for VHD building is that the build can be started from a Linux or Windows machine.

#### Kubelet and other component configuration

> Due to the lack of privileged containers in Windows, a combination of `PreKubeadmCommands`/`PostKubeadmCommands` 

Reviewer 2: Can you expand on why PreKubeadmCommands/PostKubeadmCommands are needed here? Why not put everything in the bootstrap script?

Reviewer 2: nvm just read the end of the paragraph that talks about this

> scripts and wins.exe can be used to configure the nodes. Wins.exe is currently provided as a way to  bootstrap nodes along with kubeadm in the [Kubernetes documentation for adding Windows nodes](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/adding-windows-nodes/).  

Reviewer 2: ```suggestion
scripts and wins.exe can be used to configure the nodes. Wins.exe is currently provided as a way to bootstrap nodes along with kubeadm in the [Kubernetes documentation for adding Windows nodes](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/adding-windows-nodes/).  
```

The components from the [preparenode script](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/adding-windows-nodes/#joining-a-windows-worker-node) can be used during VHD creation.

> In the future with the introduction of the [Privileged Containers for Windows containers](https://docs.google.com/document/d/12EUtMdWFxhTCfFrqhlBGWV70MkZZPOgxw0X-LTR0VAo/edit#) we can use Privileged containers in place of wins.exe enabled containers. 

Reviewer 3: Maybe link the enhancement issue here (which links to the doc) instead of the doc directly since the doc is temporary.

As the privileged container support is still being discussed, the replacement of the `PreKubeadmCommands`/`PostKubeadmCommands` kubeadm scripts might be possible but will need to be addressed at a later point in time. 

> The plan during Alpha implementation is that each of the infrastructure providers will provide their own `PreKubeadmCommands`/`PostKubeadmCommands` scripts. 

Reviewer 2: Do we expect these to be very different? why not go for a common script right away?

Author: Depending on components such as runtime/cni, different settings/configurations might be required.  I don't have a clear enough picture across the infrastructure providers to make a determination currently and didn't want to hold back alpha support.  

A script that will be useful is a [windows automatic retry based on CAPH](https://github.com/microsoft/cluster-api-provider-azurestackhci/blob/e3fc20c8d835a2b854a1c05ade2f8bf91ea4db53/templates/flavors/base/cluster-template.yaml#L65) and I will update the CEAP to add that as alpha feature that should be added.

For Beta we should be able to identify common overlapping features that can be added directly to the kubeadm cluster provisioning 
(CABPK) in a similar way that the [kubeadm-bootstrap-script.sh](https://github.com/kubernetes-sigs/cluster-api/blob/master/bootstrap/kubeadm/internal/cloudinit/kubeadm-bootstrap-script.sh) is done today.

#### netbios names

> Netbios name resolution concerns were addressed in https://github.com/kubernetes-sigs/cluster-api/issues/2217#issuecomment-598999436

Reviewer 3: Can you sumamrize outcome here so people skimming this doc don't need to follow the link?

### Infrastructure provider implementation
By leveraging cloudbase-init, an infrastructure provider implementation will require only a few changes which include:

> - Make changes to their provider api to enable Windows OS worker nodes 

Reviewer 2: ```suggestion
- Make changes to their provider api to enable Windows OS infra machines
```
?

> (CABPK) in a similar way that the [kubeadm-bootstrap-script.sh](https://github.com/kubernetes-sigs/cluster-api/blob/master/bootstrap/kubeadm/internal/cloudinit/kubeadm-bootstrap-script.sh) is done today.

Author: From the conversation https://github.com/kubernetes-sigs/cluster-api/pull/3616#discussion_r492860703, I think we should add this support in Alpha, I am going to add a new section that outlines the work required along with testing

> - Make changes to their provider api to enable Windows OS infra machines

Reviewer 5: Let's provide an example?

- Ensuring the cloudbase-init data lands in the correct place on the machines for Windows OS worker nodes

From the infrastructure provider perspective, there are no known required changes to the CAPI API to support Windows

> nodes at this time.  If during alpha we identify changes we will open issues pertaining to the changes required.

Reviewer 5: ```suggestion
nodes at this time. If during alpha we identify changes we will open issues pertaining to the changes required.
```

### User Stories

#### As an operator, I would like to create Windows OS worker nodes with the CAPI API.

#### As an operator, I would like to manage Windows OS worker nodes with the CAPI API.

### Implementation Details/Notes/Constraints

Due to the lack of privileged containers in Windows there are two options for configuring the components such as 
kube-proxy, kubelet.  The above solution using wins is preferred because it makes the move to privileged containers 
straightforward as a drop and replace. 

While this is the best choice for the alpha and the community direction there are some infrastructure providers that may 
not be able to use wins due to signing or security concerns since wins allows the execution of any arbitrary command on 
the host. Pre/post commands can be used as an alternative with additional scripts cached on the VHD that enable the 
configuration.

#### Signing of the components.  
Some infrastructure providers will require any scripts and binaries are signed before deployment.  

> This will be managed by providing the ability to provide external scripts and binaries during the VHD building process.  We will also be discussing this further with the Image Builder repository team during the process for adding the VHDs.

Reviewer 3: ```suggestion
This will be managed by providing the ability to provide external scripts and binaries during the VHD building process.  We will also be discussing this further with the Image Builder repository team during the process for adding the VHDs.
```
```suggestion
This will be managed by providing the ability to provide external scripts and binaries during the image building process.  We will also be discussing this further with the Image Builder repository team during the process for adding the images.
```

#### Known prototypes and prior work: 
- https://github.com/adelina-t/cloudbase-init-capz-demo
- https://github.com/benmoss/kubeadm-windows/tree/master/cluster-api-aws

### Security Model

> Wins is currently [recommended way to use kubeadm](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/adding-windows-nodes/).  

Reviewer 3: ```suggestion
Wins.exe is currently the [recommended way to use kubeadm](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/adding-windows-nodes/).  
```

> Limiting access to the named pipes that are required is one way to mitigate access.   Wins is currently 

Reviewer 2: ```suggestion
Limiting access to the named pipes that are required is one way to mitigate access. Wins is currently 
```

required during and after provisioning for running kube-proxy and the CNI daemonset.  

> The security model for Privileged containers is still being discussed in KEP and is still early and will be

Reviewer 3: Maybe explicitly state security concerns are not in scope for alpha but will be for beta?

Author: I've updated the wording PLTA

Kubeadm bootstrap token should be able to use multi-part mime documents for cloudbase-init as done for [Linux in CAPA](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/28d01d064cc2e5b0286ae23b3be7203f18b00447/controllers/awsmachine_controller.go#L601). 
This will require an update to Cloudbase-init which does support [mutli-part mime documents](https://cloudbase-init.readthedocs.io/en/latest/userdata.html#multi-part-content) but is missing the [boothook](https://cloudinit.readthedocs.io/en/latest/topics/format.html?highlight=boothook#cloud-boothook) functionality.  
Support for cloudhooks can be added to cloudbase-init to meet the AWS provider requirement that 
would bring parity to Linux implementation during the Beta phase.

There is [no known requirement](https://github.com/kubernetes-sigs/cluster-api/issues/2218) for managing the Admin
Kubeconfig or domain passwords in the Windows configuration. Domain passwords should be managed outside the scope of 
CAPI and only Kubeadm bootstrap tokens, which have limited lifetime, should be used for the joining the Windows nodes. 

### Risks and Mitigations

- Privileged containers are not implemented.

>   - There is an active discussion and [KEP](https://docs.google.com/document/d/12EUtMdWFxhTCfFrqhlBGWV70MkZZPOgxw0X-LTR0VAo/edit#) In place.  At the Beta stage the community can do a checkpoint to determine if the solution fits customer needs

Reviewer 2: ```suggestion
  - There is an active discussion and [KEP](https://docs.google.com/document/d/12EUtMdWFxhTCfFrqhlBGWV70MkZZPOgxw0X-LTR0VAo/edit#) in place.  At the Beta stage, the community can do a checkpoint to determine if the solution fits customer needs
```

- Cloudbase-init is a third party dependency
  - This project is under Apache 2.0 License : https://github.com/cloudbase/cloudbase-init which is cleared under the CNCF Allow list: https://github.com/cncf/foundation/blob/master/allowed-third-party-license-policy.md
- Windows VHD Distribution

>   - Cluster API already provides the ability to use customer provided VHD and images provided are only for testing demonstration purposes and it is recommended the customer creates their own image. 

Reviewer 2: This is actually infrastructure provider specific

Author: updated to reflect that it is the infra providers that can provide this ability

   containers in Beta, this becomes a non issue as wins will no longer be required.

1. Create a separate bootstrap provider for Windows.  This would require re-implementing a lot of the logic that is 
   already in CABPK.  
   When bugs are fixed or changes in behavior occur, Windows would risk being out of sync with the Linux implementation.

1. Modified CABPK provider to have a different output format than cloud-init for Windows nodes.   With Cloudbase-init 
   there are no requirements to change CABPK which makes the adaption for Windows straightforward. If we were to adapt 
   the output format of CABPK there is a potential for introduction of bugs and variation in the logic that would be 
   created for Windows nodes.  This would cause the Windows implementation to differ from others which could lead to 
   confusion when debugging differences.

## Upgrade Strategy

Nodes that use this pattern will require the infrastructure to be immutable as specified by CAPI documentation.

## Additional Details

### Test Plan [optional]

**Note:** *Section not required until targeted at a release.*

> Given there are no changes to the CAPI codebase, The testing plan would be left up to each infrastructure provider. 

Reviewer 2: "Given there are no changes to the CAPI codebase" that's not entirely true since changes are required for bootstrap kubeadm provider which is part of the CAPI codebase. How are we planning on testing that? I'm especially curious about testplans for CAPI e2e using the docker provider.

Author: at the time of writing, there were no planned changes to the bootstrap provider. I will update with the new learning and a plan for testing the windows retry logic.  

Does the docker provider use kind to create the workload cluster?

Author: Looked into this a little further. 

The changes that would be proposed would be to add `UseExperimentalRetryJoin` for windows deployments.  There were concerns about this feature discussed in https://github.com/kubernetes-sigs/cluster-api/pull/2763#discussion_r397306055. I don't see that tests were added for this particular feature.  

It looks like the docker provider uses [kind](https://kind.sigs.k8s.io/).  Windows containers do not support docker in docker so we would not be able to reuse those tests even in the case that there were tests. I think we could add unit tests to ensure the correct script is output. 

It looks from the https://github.com/kubernetes-sigs/cluster-api/issues/2769#issuecomment-603989176 that there will be a need for retries in some fashion in CABPK.  I didn't really get a sense of the future direction for retries in cabpk, has there been further discussion on the future of `UseExperimentalRetryJoin`?

Reviewer 6: I might start reviving the idea of a node agent that imports kubeadm as a library, which is a thing tat is happening and may call a meeting next week if people are interested. I think in the meantime, you could use a windows specific UseExperimentalRetryJoin. In which case cabpk would need to know if it's Linux or Windows the script is being generated for.

Author: > In which case cabpk would need to know if it's Linux or Windows the script is being generated for.

Spoke briefly with @CecileRobertMichon about where the best place for that to land OS information and came up with a few options.  It sounds like OS info might be needed for the [sentinel file in the bootstrap failure proposal](https://docs.google.com/document/d/1U0GxvO6ltgIINMjpQz96UD4bExlN2h21wyyn-3ENezc/edit) as well.

1. Add an additional OS field on the kubeadm spec next to [UseExperimentalRetryJoin](https://github.com/jsturtevant/cluster-api/blob/28b1087a10015beb421b51cbd4eade9653744710/bootstrap/kubeadm/api/v1alpha3/kubeadmbootstrapconfig_types.go#L97)
2. Add as Part of the contract for Machine spec and get it the same [k8s version is acquired](https://github.com/kubernetes-sigs/cluster-api/blob/fffcfedb09dab668b45b71c748f6609a53baa518/bootstrap/kubeadm/controllers/kubeadmconfig_controller.go#L806).

Option 2, feel more appropriate to me as the OS is a machine level construct and it could be reused across providers and other machinery. For Backwards compatibility and providers that wouldn't set it, it could default to Linux.

Reviewer 6: 2 seems fine to me since cabpk already introspects the Machine object today.

Reviewer 7: I agree with @naadir that it seems right to put it on the Machine, but also that I think ideally we'll not need OS-specific kubeadm retry behavior. Inevitably there will be other things that need to switch on OS which is why it still makes sense to put it on the Machine so that it's not a one-off fix just for this problem.

Author: > ideally we'll not need OS-specific kubeadm retry behavior

would this be resolved via the "kubeadm as a library" idea?

Reviewer 5: Just some clarifications

> kubeadm as a library, which is a thing tat is happening

This is being discussed with the kubeadm community, although I wouldn't expect anything to materialize before 2021.

> Spoke briefly with @CecileRobertMichon about where the best place for that to land OS information and came up with a few options. It sounds like OS info might be needed for the sentinel file in the bootstrap failure proposal as well.

We were talking for a different effort (https://github.com/kubernetes-sigs/cluster-api/pull/3668#issuecomment-696211676) to add a condition to the InfrastructureMachine that communicates "ready to receive bootstrap data".

The reason I'm bringing this up is that if in v1alpha4 we go down that route, we could have the infrastructure machine to optionally expose the operating system information based on the image provided. This information could be useful also for debugging purposes, or to write other code that's OS-dependent. While I just said "OS", we could have a well-defined struct with generic information that we want to expose.

Reviewer 6: > would this be resolved via the "kubeadm as a library" idea?

Potentially. There's a whole bunch of use cases not just related to Windows. Will start jotting down together and convene a WG for working parties. This shouldn't block your initial implementation though.

Author: @vincepri is there a place to add for a note for an OS/generic struct to the infra machine for v1alpha4?

I believe it would it be adding a struct/field to: https://github.com/kubernetes-sigs/cluster-api/blob/b0236f2b3c1d0ee9e6a8dffd1baaa22fd8aee17d/api/v1alpha3/machine_types.go#L57

Reviewer 2: IMO we should add the OS info to the Infra machine (AWSMachine, AzureMachine, DockerMachine, etc) since it's tightly coupled with the infrastructure and the OS image. Maybe it should be part of the infra provider contract?

Then we sync the InfraMachine with a Status field in the core CAPI Machine type, and use that in CABPK and any other bootstrap providers that want to support Windows.

Reviewer 5: Yes that sounds good to me as well, I was hoping we'd get there. CAPI's Machine object should be agnostic and look as much as possible as a Kubernetes node, rather than a VM, which is the role of the infrastructure side of things

Author: I've add some wording to capture this in a distinct commit: https://github.com/kubernetes-sigs/cluster-api/pull/3616/commits/6ac0cb2a99930555d3e4908a39501aa7d1090aaa PTAL

Reviewer 2: @rsmitty FYI since this would involve some changes around Bootstrap/Infra provider contract

Reviewer 8: Seems fine to me. Adding OS info to the MetalMachine for our infra provider shouldn't be a big deal at all.

Author: I've updated this and added contract information in the docs PTAL

Author: Based on the conversation today at cluster-api meeting, the desire is remove experimental retry for for v1Alpha4. Since Windows support will be in v1alpha 3 I am going to drop the retry aspect. We can revisit the OsType field if a concrete use case comes up later.

It would be recommended to leverage the existing upstream Windows tests to show that Windows nodes are operating 
effectively.  If changes are required to the CAPI codebase unit tests and if appropriate e2e tests should be added.

### Graduation Criteria [optional]

**Note:** *Section not required until targeted at a release.*

#### Alpha
`PreKubeadmCommands`/`PostKubeadmCommands` command scripts are created per infrastructure provider
Initially implemented with wins and pre/post commands
Windows Packer script to create VHD with cloudbase-init added to the image builder scripts

#### Beta
Pre/Post commands moved to bootstrap provider
Adopt privileged containers (dependant on Privileged containers KEP)

> kubeadm bootstrap token can be kept secret via multi-part mime documents for cloudbase-init. 

Reviewer 9: If this is meant to be a list prepend `-`. It currently shows as a run off sentence.

  - "@ksubrmnn"
reviewers:
  - "@CecileRobertMichon"
  - "@ncdc"
  - "@randomvariable"
creation-date: 2020-08-25
last-updated: 2020-09-09
status: implementable
see-also:
---

# Windows kubeadm-based worker nodes support

## Table of Contents

- [Windows kubeadm-based worker nodes support](#windows-kubeadm-based-worker-nodes-support)
  - [Table of Contents](#table-of-contents)
  - [Glossary](#glossary)
  - [Summary](#summary)
  - [Motivation](#motivation)
    - [Goals](#goals)
    - [Non-Goals/Future Work](#non-goalsfuture-work)
  - [Proposal](#proposal)
    - [Cluster API Bootstrap Provider Kubeadm](#cluster-api-bootstrap-provider-kubeadm)
      - [cloud-init and cloudbase-init](#cloud-init-and-cloudbase-init)
      - [Image Creation](#image-creation)
      - [Kubelet and other component configuration](#kubelet-and-other-component-configuration)
      - [netbios names](#netbios-names)
    - [Infrastructure provider implementation](#infrastructure-provider-implementation)
    - [User Stories](#user-stories)
      - [As an operator, I would like to create Windows OS worker nodes with the CAPI API.](#as-an-operator-i-would-like-to-create-windows-os-worker-nodes-with-the-capi-api)

>       - [As an operator, I would like to manage Windows OS worker nodes with the CAPI API.](#as-an-operator-i-would-like-to-manage-windows-os-worker-nodes-with-the-capi-api)

Reviewer 10: Seems like some type active directory awareness would be worth diving into.  
```
- [As an operator, i'd like to federate my windows OS worker nodes with active directory]
```
or is that out of scope?  Just a thought from what we think our customers would like to be able to do

Reviewer 10: (feel free to merge if its ready , dont wanna hold things up, maybe can iterate if this is in scope)

Author: Yes definitely. I think we decided to move forward with this and address the AD via https://github.com/kubernetes-sigs/cluster-api/issues/3761

ref: https://github.com/kubernetes-sigs/cluster-api/pull/3616#issuecomment-708358305

the same configuration because [Ansible](https://docs.ansible.com/ansible/latest/user_guide/windows.html) requires [Windows specific modules](https://docs.ansible.com/ansible/latest/modules/list_of_windows_modules.html) to do the configuration. 

#### Kubelet and other component configuration

Due to the lack of privileged containers in Windows, a combination of `PreKubeadmCommands`/`PostKubeadmCommands` 
scripts and wins.exe can be used to configure the nodes. Wins.exe is currently provided as a way to bootstrap nodes along with kubeadm in the [Kubernetes documentation for adding Windows nodes](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/adding-windows-nodes/).  
The components from the [preparenode script](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/adding-windows-nodes/#joining-a-windows-worker-node) can be used during image creation.

In the future, when support for [Privileged Containers for Windows containers](https://github.com/kubernetes/enhancements/issues/1981) is merged, we might be able to revisit this proposal 
and use privileged containers in place of wins.exe enabled containers. 

> Each infrastructure providers must provide their own `PreKubeadmCommands`/`PostKubeadmCommands` scripts that

Reviewer 11: Can you elaborate on what each infrastructure provider should do? What would be a sample PreKubeadmCommand for azure? Downloading Windows base image like server-core or something like that?

The components from the [preparenode script](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/adding-windows-nodes/#joining-a-windows-worker-node) can be used during image creation.

In the future, when support for [Privileged Containers for Windows containers](https://github.com/kubernetes/enhancements/issues/1981) is merged, we might be able to revisit this proposal 
and use privileged containers in place of wins.exe enabled containers. 

Each infrastructure providers must provide their own `PreKubeadmCommands`/`PostKubeadmCommands` scripts that
are required for additional configuration for the node. During planning for Beta we will be able to identify
common overlapping features that can be added into the the base images in image-builder for re-use.

#### Kubeadm retry support 

Based on the [CAPH Windows implementation](https://github.com/microsoft/cluster-api-provider-azurestackhci), it was 

> identified that the [experimental retry](https://github.com/kubernetes-sigs/cluster-api/pull/2763) for kubeadm 

Reviewer 6: I've opened https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1993 to test if it's still needed. Should do the same for the other providers.

> identified that the [experimental retry](https://github.com/kubernetes-sigs/cluster-api/pull/2763) for kubeadm 

Author: I've added an issue for Azure: https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/998

If the retry support is sufficient, I can remove this section all together.  I still think there could be reason down the line for `osType` on the infra machine, but we can address that later if there is a solid reason.

> identified that the [experimental retry](https://github.com/kubernetes-sigs/cluster-api/pull/2763) for kubeadm 

Author: Based on the conversation today at cluster-api meeting, the desire is remove experimental retry for  for v1Alpha4. Since Windows support will be in v1alpha 3 I am going to drop the retry aspect.  We can revisit the `OsType` field if a concrete use case comes up later.

Each infrastructure providers must provide their own `PreKubeadmCommands`/`PostKubeadmCommands` scripts that
are required for additional configuration for the node. During planning for Beta we will be able to identify
common overlapping features that can be added into the the base images in image-builder for re-use.

#### Kubeadm retry support 

Based on the [CAPH Windows implementation](https://github.com/microsoft/cluster-api-provider-azurestackhci), it was 
identified that the [experimental retry](https://github.com/kubernetes-sigs/cluster-api/pull/2763) for kubeadm 
cluster provisioning (CABPK) will be beneficial for stability of Windows node joins.  The `UseExperimentalRetryJoin` 

> currently uses a Linux shell script to create add the retry functionality and does not work on Windows. The ideal 

Reviewer 7: Can we just make this script into a cross-platform Go binary that we do the retry with?

Author: In the meeting there was a concern around versioning and binary distribution.  The end goal will be using a kubeadm as a library. 

The OsType information could be used by other providers as well which adds some additional value in providing the proposal above: https://github.com/kubernetes-sigs/cluster-api/pull/3616#discussion_r495132309

Author: Retry support in kube-adm was add in 1.19 and back-ported to 1.17.  As of the meeting today the goal is to drop the ExperimentalRetry via v1alpha. Any issues that arise we should open in kubeadm moving forward
